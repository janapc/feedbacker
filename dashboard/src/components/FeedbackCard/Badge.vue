<template>
  <span id="bagde-component" :class="`bg-${classColor}`" class="p-2 text-xs font-black text-white uppercase rounded-full">{{
    label
  }}</span>
</template>

<script>
import { computed } from 'vue';

export default {
  props: {
    type: { type: String, required: true }
  },
  setup (props) {
    const label = computed(() => {
      if (props.type === 'ISSUE') {
        return 'problema';
      }

      if (props.type === 'IDEA') {
        return 'idÃ©ia';
      }

      return 'outros';
    });

    const classColor = computed(() => {
      if (props.type === 'ISSUE') {
        return 'brand-danger';
      }

      if (props.type === 'IDEA') {
        return 'brand-warning';
      }

      return 'brand-graydark';
    });

    return {
      classColor,
      label
    };
  }
};
</script>
