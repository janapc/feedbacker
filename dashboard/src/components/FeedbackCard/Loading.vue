<template>
  <content-loader class="flex flex-col items-center rounded" width="100%" heigh="300px">
    <content-loader class="mt-3 rounded" width="90%" heigh="90" animation-duration="2s" />
    <content-loader class="mt-3 rounded" width="90%" heigh="90" animation-duration="2.3s" />
    <content-loader class="mt-3 rounded" width="90%" heigh="90" animation-duration="2.7s" />
  </content-loader>
</template>

<script>
import ContentLoader from '../ContentLoader/index.vue';

export default {
  components: { ContentLoader }
};
</script>
