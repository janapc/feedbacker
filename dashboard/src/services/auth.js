export default (httpClient) => ({
  register: async ({ name, email, password }) => {
    const response = await httpClient.post('/auth/register', { name, email, password });

    let errors = null;

    if (!response.data) {
      errors = {
        status: response.request.status,
        statusText: response.request.statusText,
      };
    }

    return {
      data: response.data,
      errors,
    };
  },
  login: async ({ email, password }) => {
    const response = await httpClient.post('/auth/login', { email, password });
    let errors = null;

    if (!response.data) {
      errors = {
        status: response.request.status,
        statusText: response.request.statusText,
      };
    }

    return {
      data: response.data,
      errors,
    };
  },
});
