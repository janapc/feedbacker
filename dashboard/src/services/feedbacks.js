const defaultPagination = {
  limit: 5,
  offset: 0,
};

export default (httpClient) => ({
  getAll: async ({ type, limit, offset } = defaultPagination) => {
    const query = { limit, offset };
    if (type) {
      query.type = type;
    }
    const response = await httpClient.get('/feedbacks', { params: query });

    return { data: response.data };
  },
  getSummary: async () => {
    const response = await httpClient.get('/feedbacks/summary');

    return { data: response.data };
  },
});
