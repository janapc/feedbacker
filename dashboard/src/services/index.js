import axios from 'axios';

import router from '../router';
import AuthService from './auth';
import UsersService from './users';
import FeedbacksService from './feedbacks';

import { setGlobalLoading } from '../store/global';

const API_ENVS = {
  production: 'https://backend-vue3js-ixvpypm03.vercel.app/',
  development: 'http://localhost:3000',
  local: 'http://localhost:3000',
};

const httpClient = axios.create({
  baseURL: API_ENVS[process.env.NODE_ENV] ?? API_ENVS.local,
});

httpClient.interceptors.response.use(
  (response) => {
    setGlobalLoading(false);
    return response;
  },
  (error) => {
    const canThrowAnError = error.request.status === 0 || error.request.status === 500;

    if (canThrowAnError) {
      setGlobalLoading(false);
      throw new Error(error.message);
    }
    if (error.response.status === 401) {
      router.push({ name: 'Home' });
    }
    setGlobalLoading(false);
    return error;
  },
);

httpClient.interceptors.request.use((config) => {
  setGlobalLoading(true);
  const token = window.localStorage.getItem('token');

  if (token) {
    config.headers.common.Authorization = `Bearer ${token}`;
  }
  return config;
});

export default {
  auth: AuthService(httpClient),
  users: UsersService(httpClient),
  feedbacks: FeedbacksService(httpClient),
};
