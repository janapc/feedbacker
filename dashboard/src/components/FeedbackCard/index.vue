<template>
  <div id="feedback-card" @click="handleToggle" class="flex flex-col px-8 py-6 rounded cursor-pointer bg-brand-gray">
    <div class="flex items-center justify-between w-full mb-8">
      <badge :type="feedback.type" id="field-badge"/>

      <span class="font-regular text-brand-graydark" id="field-createdAt">
        {{ getDiffTimeBetweenCurrentDate(feedback.createdAt) }}
      </span>
    </div>

    <div class="text-lg font-medium text-gray-800" id="field-text">
      {{ feedback.text }}
    </div>

    <div
      :class="{
        animate__fadeOutUp: state.isClosing
      }"
      class="flex mt-8 animate__animated animate__fadeInUp animate__faster"
      v-if="state.isOpen"
      id="open-card"
    >
      <div class="flex flex-col w-1/2">
        <div class="flex flex-col">
          <span class="font-bold text-gray-400 uppercase select-none">Página</span>
          <span class="font-medium text-gray-700" id="field-page">{{ feedback.page }}</span>
        </div>

        <div class="flex flex-col">
          <span class="font-bold text-gray-400 uppercase select-none">Dispositivo</span>
          <span class="font-medium text-gray-700" id="field-device">{{ feedback.device }}</span>
        </div>
      </div>

      <div class="flex flex-col w-1/2">
        <div class="flex flex-col">
          <span class="font-bold text-gray-400 uppercase select-none">Usuário</span>
          <span class="font-medium text-gray-700" id="field-fingerprint">{{ feedback.fingerprint }}</span>
        </div>
      </div>
    </div>

    <div class="flex justify-end mt-8" v-if="!state.isOpen" id="close-chevron">
      <icon name="chevron-down" size="24" :color="brandColors.graydark" />
    </div>
  </div>
</template>

<script>
import { reactive } from 'vue';
import Badge from './Badge.vue';
import getDiffTimeBetweenCurrentDate from '../../utils/date';
import palette from '../../../palette';
import Icon from '../Icon/index.vue';
import wait from '../../utils/timeout';

export default {
  components: { Badge, Icon },
  props: {
    feedback: { type: Object, required: true },
    isOpened: { type: Boolean, default: false }
  },
  setup (props) {
    const state = reactive({
      isOpen: props.isOpened,
      isClosing: !props.isOpened
    });

    async function handleToggle () {
      state.isClosing = true;

      await wait(250);
      state.isOpen = !state.isOpen;
      state.isClosing = false;
    }

    return {
      handleToggle,
      getDiffTimeBetweenCurrentDate,
      brandColors: palette.brand,
      state
    };
  }
};
</script>
