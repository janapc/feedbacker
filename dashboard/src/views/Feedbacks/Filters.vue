<template>
  <div class="flex flex-col">
    <h1 class="text-2xl font-regular text-brand-darkgray">Filtros</h1>

    <ul class="flex flex-col mt-3 list-none">
      <li
        v-for="filter in state.filters"
        :key="filter.label"
        :id="`${filter.label}-filter`"
        :class="{
          'bg-gray-200 bg-opacity-50': filter.active
        }"
        @click="() => handleSelect(filter)"
        class="flex items-center justify-between px-4 py-1 rounded cursor-pointer"
      >
        <div class="flex items-center">
          <span :class="filter.color.bg" class="inline-block w-2 h-2 mr-2 rounded-full" />
          {{ filter.label }}
        </div>
        <span
          :class="filter.active ? filter.color.text : 'text-brand-graydark'"
          class="font-bold"
          :id="`${filter.label}-filter-amount`"
          >{{ filter.amount }}</span
        >
      </li>
    </ul>
  </div>
</template>

<script>
import { reactive } from 'vue';
import services from '../../services';
import useStore from '../../hooks/useStore';

const LABELS = {
  all: 'Todos',
  issue: 'Problemas',
  idea: 'Ideais',
  other: 'Outros',
};

const COLORS = {
  all: { text: 'text-brand-info', bg: 'bg-brand-info' },
  issue: { text: 'text-brand-danger', bg: 'bg-brand-danger' },
  idea: { text: 'text-brand-warning', bg: 'bg-brand-warning' },
  other: { text: 'text-brand-graydark', bg: 'bg-brand-graydark' },
};

function applyFiltersStructure(summary) {
  return Object.keys(summary).reduce((acc, cur) => {
    const currentFilter = {
      label: LABELS[cur],
      color: COLORS[cur],
      amount: summary[cur],
    };

    if (cur === 'all') {
      currentFilter.active = true;
    } else {
      currentFilter.type = cur;
    }

    return [...acc, currentFilter];
  }, []);
}

export default {
  async setup(props, { emit }) {
    const store = useStore('Global');
    const state = reactive({
      hasError: false,
      filters: [{ label: null, amount: null }],
    });

    try {
      const { data } = await services.feedbacks.getSummary();
      state.filters = applyFiltersStructure(data);
    } catch (error) {
      state.hasError = !!error;
      state.filters = applyFiltersStructure({
        all: 0,
        issue: 0,
        idea: 0,
        other: 0,
      });
    }

    function handleSelect({ type }) {
      if (store.isLoading) {
        return;
      }

      state.filters = state.filters.map((filter) => {
        if (filter.type === type) {
          return { ...filter, active: true };
        }
        return { ...filter, active: false };
      });

      emit('select', type);
    }

    return {
      state,
      handleSelect,
    };
  },
};
</script>
