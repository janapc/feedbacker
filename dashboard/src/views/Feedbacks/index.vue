<template>
  <div class="flex justify-center w-full h-28 bg-brand-main"><header-logged /></div>

  <div class="flex flex-col items-center justify-center h-64 bg-brand-gray">
    <h1 class="text-4xl font-black text-center text-gray-800">Feedbacks</h1>
    <p class="text-lg text-center text-gray-800 font-regular">
      Detalhes de todos os feedbacks recebidos.
    </p>
  </div>

  <div class="flex justify-center w-full pb-20">
    <div class="w-4/5 max-w-6xl py-10 grid grid-cols-4 gap-2">
      <div>
        <h1 class="text-3xl font-black text-brand-darkgray">Listagem</h1>

        <suspense>
          <template #default
            ><filters
              @select="changeFeedbacksType"
              class="mt-8 animate__animated animate__fadeIn animate__faster"
          /></template>
          <template #fallback><filters-loading class="mt-8"/></template>
        </suspense>
      </div>

      <div class="px-10 pt-20 col-span-3" id="list-of-feedbacks">
        <p
          id="feedbacks-empty-message"
          v-if="
            !state.feedbacks.length &&
              !state.isLoading &&
              !state.isLoadingFeedbacks &&
              !state.hasError
          "
          class="text-lg text-center text-gray-800 font-regular"
        >
          Ainda nenhum feedback recebido ðŸ¤“
        </p>
        <p
          id="hasError-message"
          v-if="state.hasError"
          class="text-lg text-center text-gray-800 font-regular"
        >
          Aconteceu um erro ao carregar os feedbacks ðŸ˜”
        </p>

        <feedback-card-loading v-if="state.isLoading || state.isLoadingFeedbacks" />

        <feedback-card
          v-else
          v-for="(feedback, index) in state.feedbacks"
          :key="feedback.id"
          :id="`${index}-card`"
          :is-opened="index === 0"
          :feedback="feedback"
          class="mb-8 feedbacks-cards"
        />

        <feedback-card-loading v-if="state.isLoadingMoreFeedbacks" />
      </div>
    </div>
  </div>
</template>

<script>
import {
  onErrorCaptured, onMounted, onUnmounted, reactive,
} from 'vue';
import HeaderLogged from '../../components/HeaderLogged/index.vue';
import FeedbackCard from '../../components/FeedbackCard/index.vue';
import FeedbackCardLoading from '../../components/FeedbackCard/Loading.vue';
import Filters from './Filters.vue';
import FiltersLoading from './FiltersLoading.vue';
import services from '../../services';

export default {
  components: {
    HeaderLogged,
    Filters,
    FiltersLoading,
    FeedbackCard,
    FeedbackCardLoading,
  },
  setup() {
    const state = reactive({
      isLoading: false,
      isLoadingFeedbacks: false,
      isLoadingMoreFeedbacks: false,
      hasError: false,
      feedbacks: [],
      currentFeedbackType: '',
      pagination: {
        limit: 5,
        offset: 0,
      },
    });

    function handleErros(error) {
      state.isLoading = false;
      state.isLoadingFeedbacks = false;
      state.isLoadingMoreFeedbacks = false;
      state.hasError = !!error;
    }

    async function changeFeedbacksType(type) {
      try {
        state.isLoadingFeedbacks = true;
        state.pagination.limit = 5;
        state.pagination.offset = 0;
        state.currentFeedbackType = type;

        const { data } = await services.feedbacks.getAll({
          type,
          ...state.pagination,
        });

        state.feedbacks = data.results;
        state.pagination = data.pagination;
        state.isLoadingFeedbacks = false;
      } catch (error) {
        handleErros(error);
      }
    }

    async function fetchFeedbacks() {
      try {
        state.isLoading = true;

        const { data } = await services.feedbacks.getAll({
          ...state.pagination,
          type: state.currentFeedbackType,
        });

        state.feedbacks = data.results;
        state.pagination = data.pagination;
        state.isLoading = false;
      } catch (error) {
        handleErros(error);
      }
    }

    async function handleScroll() {
      const isBottomOfWindow = Math.ceil(document.documentElement.scrollTop + window.innerHeight)
        >= document.documentElement.scrollHeight;

      if (state.isLoading || state.isLoadingMoreFeedbacks) return;
      if (!isBottomOfWindow) return;
      if (state.feedbacks.length >= state.pagination.total) return;

      try {
        state.isLoadingMoreFeedbacks = true;
        const { data } = await services.feedbacks.getAll({
          ...state.pagination,
          type: state.currentFeedbackType,
          offset: state.pagination.offset + 5,
        });

        if (data.results.length) {
          state.feedbacks.push(...data.results);
        }

        state.pagination = data.pagination;
        state.isLoadingMoreFeedbacks = false;
      } catch (error) {
        handleErros(error);
      }
    }

    onMounted(() => {
      fetchFeedbacks();
      window.addEventListener('scroll', handleScroll, false);
    });

    onUnmounted(() => {
      window.removeEventListener('scroll', handleScroll, false);
    });

    onErrorCaptured(handleErros);

    return { state, changeFeedbacksType };
  },
};
</script>

<style></style>
