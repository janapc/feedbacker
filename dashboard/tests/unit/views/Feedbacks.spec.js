import { flushPromises, mount, shallowMount } from '@vue/test-utils';

import Feedbacks from '../../../src/views/Feedbacks/index.vue';
import FeedbackCard from '../../../src/components/FeedbackCard/index.vue';
import Filters from '../../../src/views/Feedbacks/Filters.vue';

import services from '../../../src/services';
import wait from '../../../src/utils/timeout';

import mockData from './getFeedbacks.json';

jest.mock('../../../src/services');

describe('<Feedbacks />', () => {
  afterEach(() => {
    jest.resetAllMocks();
    jest.clearAllMocks();
    jest.restoreAllMocks();
  });

  it('should render Feedbacks correctly', () => {
    services.feedbacks.getAll.mockReturnValue(mockData.all);

    const wrapper = shallowMount(Feedbacks);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Should show all the feedbacks', async () => {
    services.feedbacks.getAll.mockReturnValue(mockData.all);

    const wrapper = shallowMount(Feedbacks);

    await wait(500);

    expect(services.feedbacks.getAll).toBeCalledWith({ limit: 5, offset: 0, type: '' });

    expect(wrapper.findAllComponents(FeedbackCard).length).toEqual(2);
  });

  it('Should show only a filter specific', async () => {
    services.feedbacks.getAll
      .mockReturnValueOnce(mockData.all)
      .mockReturnValueOnce(mockData.filterSpecific);
    services.feedbacks.getSummary.mockReturnValueOnce({
      all: 2,
      issue: 1,
      idea: 0,
      other: 1
    });

    const wrapper = mount(Feedbacks);

    await wait(500);

    expect(wrapper.findComponent(Filters).exists()).toBe(true);
    expect(wrapper.findAllComponents(FeedbackCard).length).toEqual(2);
    expect(services.feedbacks.getAll).toBeCalledWith({ limit: 5, offset: 0, type: '' });
    expect(services.feedbacks.getSummary).toBeCalled();

    await wrapper.find('#Problemas-filter').trigger('click');
    await flushPromises();

    expect(services.feedbacks.getAll).toBeCalledWith({ limit: 5, offset: 0, type: 'issue' });
    expect(wrapper.findAllComponents(FeedbackCard).length).toEqual(1);
  });

  it('Should show more feedbacks', async () => {
    services.feedbacks.getAll
      .mockReturnValueOnce(mockData.all)
      .mockReturnValueOnce(mockData.newPage);

    const wrapper = shallowMount(Feedbacks);

    await wait(500);

    window.dispatchEvent(new Event('scroll'));

    expect(services.feedbacks.getAll).toBeCalledWith({ limit: 5, offset: 0, type: '' });
    expect(services.feedbacks.getAll).toBeCalledWith({ limit: 5, offset: 0, type: '' });

    await wait(500);

    expect(wrapper.findAllComponents(FeedbackCard).length).toEqual(3);
  });

  it('Should return a message error if not have the feedbacks', async () => {
    services.feedbacks.getAll.mockReturnValueOnce({
      data: {
        results: [],
        pagination: {
          limit: 5,
          offset: 0
        }
      }
    });
    const wrapper = shallowMount(Feedbacks);

    await wait(500);

    expect(wrapper.find('#feedbacks-empty-message').text()).toEqual(
      'Ainda nenhum feedback recebido ðŸ¤“'
    );
  });

  it('Should return a message error if not load the feedbacks', async () => {
    services.feedbacks.getAll.mockReturnValueOnce({});

    const wrapper = shallowMount(Feedbacks);

    await wait(500);

    expect(wrapper.find('#hasError-message').text()).toEqual(
      'Aconteceu um erro ao carregar os feedbacks ðŸ˜”'
    );
  });
});
