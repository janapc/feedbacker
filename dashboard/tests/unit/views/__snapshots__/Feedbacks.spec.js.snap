// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Feedbacks /> should render Feedbacks correctly 1`] = `
<div class="flex justify-center w-full h-28 bg-brand-main">
  <header-logged-stub></header-logged-stub>
</div>
<div class="flex flex-col items-center justify-center h-64 bg-brand-gray">
  <h1 class="text-4xl font-black text-center text-gray-800">Feedbacks</h1>
  <p class="text-lg text-center text-gray-800 font-regular"> Detalhes de todos os feedbacks recebidos. </p>
</div>
<div class="flex justify-center w-full pb-20">
  <div class="w-4/5 max-w-6xl py-10 grid grid-cols-4 gap-2">
    <div>
      <h1 class="text-3xl font-black text-brand-darkgray">Listagem</h1>
      <stub></stub>
    </div>
    <div class="px-10 pt-20 col-span-3" id="list-of-feedbacks">
      <p id="feedbacks-empty-message" class="text-lg text-center text-gray-800 font-regular"> Ainda nenhum feedback recebido ðŸ¤“ </p>
      <!--v-if-->
      <!--v-if-->
    </div>
  </div>
</div>
`;
